File: 828 Deciphering Messages/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <stdbool.h>

char deciph(char ch, int N) {
    return (ch - 'A' + 26 - N) % 26 + 'A';
}

int main() {
    int cases;
    scanf("%d", &cases);
    char *buffer = NULL;
    size_t bufsize = 0;
    while(cases--) {
        char *alphaKey = NULL;
        int numericalKey, messages;
        scanf("%ms %d %d", &alphaKey, &numericalKey, &messages);
        bool L[26] = {};
        int alphaKeyLen = strlen(alphaKey);
        for(int i = 0; i < alphaKeyLen; ++i)
            L[alphaKey[i] - 'A'] = true;
        getchar();
        while(messages--) {
            getline(&buffer, &bufsize, stdin);
            char *plaintext = NULL;
            int size = 0, m = 0, curr = 0, len = strlen(buffer);
            bool error = false;
            while(curr < len && !error) {
                if(!isupper(buffer[curr])) {
                    plaintext = (char*)realloc(plaintext, (size + 2) * sizeof(char));
                    plaintext[size++] = buffer[curr++];
                } else {
                    if(curr < len - 2 && isupper(buffer[curr + 1]) && isupper(buffer[curr + 2])) {
                        if(buffer[curr] == alphaKey[m] && buffer[curr + 2] == alphaKey[(m + 1) % alphaKeyLen]) {
                            char deciphChar = deciph(buffer[curr + 1], numericalKey);
                            if(L[deciphChar - 'A']) {
                                plaintext = (char*)realloc(plaintext, (size + 2) * sizeof(char));
                                plaintext[size++] = deciphChar;
                                ++m %= alphaKeyLen;
                                curr += 3;
                            } else
                                error = true;
                        } else {
                            char deciphChar = deciph(buffer[curr], numericalKey);
                            if(!L[deciphChar - 'A']) {
                                plaintext = (char*)realloc(plaintext, (size + 2) * sizeof(char));
                                plaintext[size++] = deciphChar;
                                ++curr;
                            } else
                                error = true;
                        }
                    } else {
                        char deciphChar = deciph(buffer[curr], numericalKey);
                        if(!L[deciphChar - 'A']) {
                            plaintext = (char*)realloc(plaintext, (size + 2) * sizeof(char));
                            plaintext[size++] = deciphChar;
                            ++curr;
                        } else
                            error = true;
                    }
                }
            }
            if(!error) {
                plaintext[size] = '\0';
                fputs(plaintext, stdout);
            } else
                puts("error in encryption");
            free(plaintext); 
        }
        free(alphaKey);
        if(cases)
            putchar('\n');
    }
    free(buffer);
    return 0;
}

File: 828 Deciphering Messages/solution.java
import java.io.*;

public class Main {
	public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StreamTokenizer st = new StreamTokenizer(br);
        st.nextToken();
        int cases = (int)st.nval;
        StringBuilder output = new StringBuilder();
        while(cases-- > 0) {
            st.nextToken();
            String alphaKey = st.sval;
            st.nextToken();
            int numericalKey = (int)st.nval;
            st.nextToken();
            int messages = (int)st.nval;
            boolean[] L = new boolean[26];
            int alphaKeyLen = ;
            for(int i = 0; i < alphaKey.length(); ++i)
                L[alphaKey.charAt(i) - 'A'] = true;
            while(messages-- > 0) {
                String cipheredtext = br.readLine();
                StringBuilder plaintext = new StringBuilder();
                int m = 0, curr = 0, len = cipheredtext.length();
                boolean error = false;
                while(curr < len && !error) {
                    if(!Character.isUpperCase(cipheredtext.charAt(curr)))
                        plaintext.append(cipheredtext.charAt(curr++));
                    else {
                        if(curr < len - 2 && Character.isUpperCase(cipheredtext.charAt(curr + 1)) && Character.isUpperCase(cipheredtext.charAt(curr + 2))) {
                            if(cipheredtext.charAt(curr) == alphaKey.charAt(m) && cipheredtext.charAt(curr + 2) == alphaKey.charAt((m + 1) % alphaKeyLen)) {
                                char deciphChar = deciph(cipheredtext.charAt(curr + 1), numericalKey);
                                if(L[deciphChar - 'A']) {
                                    plaintext.append(deciphChar);
                                    m = ++m % alphaKey.length();
                                    curr += 3;
                                } else
                                    error = true;
                            } else {
                                char deciphChar = deciph(cipheredtext.charAt(curr), numericalKey);
                                if(!L[deciphChar - 'A']) {
                                    plaintext.append(deciphChar);
                                    ++curr;
                                } else
                                    error = true;
                            }
                        } else {
                            char deciphChar = deciph(cipheredtext.charAt(curr), numericalKey);
                            if(!L[deciphChar - 'A']) {
                                plaintext.append(deciphChar);
                                ++curr;
                            } else
                                error = true;
                        }
                    }
                }
                if(!error)
                    output.append(plaintext).append('\n');
                else
                    output.append("error in encryption\n");
            }
            if(cases > 0)
                output.append('\n');
        }
	    System.out.print(output);
	}
	
	static char deciph(char ch, int N) {
	    return (char)((ch - 'A' + 26 - N) % 26 + 'A');
	}
}
